cobc/dump_tree.c

```diff
@@ -0,0 +1,568 @@
+
+	fmt.format_begin_flags = "[";
+	fmt.format_begin_flag = "\"";
+	fmt.format_end_flag = "\",";
+	fmt.format_end_flags = " NULL]";
+
+	fmt.format_null = "null";
+	fmt.format_begin_constr = "{ \"type_\": \"";
+	fmt.format_delim_constr = "\", \"value_\": ";
+	fmt.format_end_constr = " }";
+	fmt.format_begin_pointer = "\"";
+	fmt.format_end_pointer = "\"";
+	fmt.format_begin_char =  "\"";
+	fmt.format_end_char = "\"";
+	fmt.format_begin_string =  "\"";
+	fmt.format_end_string = "\"";
+	fmt.format_begin_int = "";
+	fmt.format_end_int = "";
+}
+
+static
+void set_format_by_file_ext (const char* filename)
+{
+	const char *arg = filename + strlen(filename);
+
+	while ( arg > filename && *arg != '.' ) arg--;
+
+	if (!strcasecmp(arg, ".ml")){
+
+		set_ocaml_format ();
+
+		cb_dump_tree_with_common = 1;
+		cb_dump_tree_with_loc = 1;
+		cb_dump_tree_with_type = 1;
+		cb_dump_tree_with_pointer = 1;
+		cb_dump_tree_with_indent = 1;
+		cb_dump_tree_with_newlines = 1;
+
+		return;
+	}
+
+	set_json_format ();
+}
+
+static
+void set_flags (const char* flags)
+{
+	int sign = 1;
+	int i;
+	int len = strlen(flags);
+
+	for(i=0; i< len; i++){
+		switch(flags[i]){
+		case '+': sign = 1; break;
+		case '-': sign = 0; break;
+		case 'c':
+			cb_dump_tree_with_common = sign;
+			break;
+		case 'l':
+			cb_dump_tree_with_loc = sign;
+			if (cb_dump_tree_with_loc){
+				cb_dump_tree_with_common = 1;
+			}
+			break;
+		case 't':
+			cb_dump_tree_with_type = sign;
+			break;
+		case 'p':
+			cb_dump_tree_with_pointer = sign;
+			break;
+		case 'i':
+			cb_dump_tree_with_indent = sign;
+			if (cb_dump_tree_with_indent)
+				cb_dump_tree_with_newlines = 1;
+			break;
+		case 'n':
+			cb_dump_tree_with_newlines = sign;
+			if (!cb_dump_tree_with_newlines)
+				cb_dump_tree_with_indent = 0;
+			break;
+		case 'm':
+			cb_dump_tree_print_message = sign;
+		case 'J':
+			set_json_format ();
+			break;
+		case 'O':
+			set_ocaml_format ();
+			break;
+		case 'A':
+			cb_dump_tree_with_loc = sign;
+			cb_dump_tree_with_common = sign;
+			cb_dump_tree_with_type = sign;
+			cb_dump_tree_with_pointer = sign;
+			cb_dump_tree_with_indent = 1-sign;
+			cb_dump_tree_with_newlines = 1-sign;
+			break;
+		default:
+			cobc_err_exit ("--dump-tree-flags: unknown flag '%c', expecting '+-cltpinA'", flags[i]);
+		}
+	}
+}
+
+void cb_dump_tree_to_file (struct cb_program *prog, const char *filename, const char* flags)
+{
+	char *env;
+	int close_fd = 1;
+
+	set_format_by_file_ext (filename);
+	env = getenv ("COB_DUMP_TREE_FLAGS");
+	if (env) set_flags (env);
+	if (flags) set_flags (flags);
+
+	if (!strcmp(filename, COB_DASH) || !strcmp(filename, "stdout")){
+		close_fd = 0;
+		fd = stdout;
+	} else
+		if (!strcmp(filename, "stderr")){
+			close_fd = 0;
+			fd = stderr;
+		} else {
+			int len = strlen (filename);
+			struct stat st;
+			if (len > 0 && filename[len-1] == '/'){
+				filename = cobc_main_stradd_dup (filename, prog->program_id);
+			}
+			if (!stat (filename, &st)) unlink(filename);
```

Comment:
Why the unlink? fopen as specified will overwrite this, no?
And if using `cobc prog1.cob prog2.cop -fdump-file=progs.dump` we'll have only one element included that way, no?

Just as a question: would the dump for multiple files in one be bad? If not allowing starting the filename with a "+" (adding that to the mode) would be good.

In any case: please add the check for the unix_lf (binary mode) here to not break the upcoming tests under Windows.
What about one source with multiple programs (contained and or not contained)?

Reply:
This one is open (why the unlink, what happens with multiple programs in the command line, addition of optional "+" and the unix-lf...)